oshmem/shmem/c/shmem_mcs_lock.c

```diff
@@ -0,0 +1,256 @@
+		 *	This value to be changed eventually by predecessor
+		 *  when its lock is released.
+		 *  Need to be done atomically to avoid any races where
+		 *  next pointer is modified by another PE acquiring or
+		 *  releasing this.
+		 **/
+		retv = MCA_ATOMIC_CALL(add(oshmem_ctx_default, (void *)next, SIGNAL_MASK,
+						sizeof(int), my_pe));
+		RUNTIME_CHECK_RC(retv);
+		MCA_SPML_CALL(quiet(oshmem_ctx_default));
+		/**
+		 * Wait for predecessor release lock to this PE
+		 * signal to false.
+		 */
+		retv = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default, (void*)next,
+				(void*)&curr, tval, sizeof(int), my_pe));
+		RUNTIME_CHECK_RC(retv);
+		while (SIGNAL(curr)) {
+			retv = MCA_SPML_CALL(wait((void*)next, SHMEM_CMP_NE, (void*)&curr,
+						SHMEM_INT));
+		    RUNTIME_CHECK_RC(retv);
+		    retv = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default, (void*)next,
+					(void*)&curr, tval, sizeof(int), my_pe));
+		    RUNTIME_CHECK_RC(retv);
+		}
+	}
+	/** else.. this pe has got the lock as no one else had it */
+}
+
+void
+_shmem_mcs_clear_lock(long *lockp)
+{
+	lock_t	*lock = (lock_t *) lockp;
+	int	mcs_tail_owner = MCSQ_TAIL_OWNER(lock);
+	int	*tail = &(lock->tail);
+	int	*next = &(lock->next);
+	int	my_pe = shmem_my_pe();
+	int	next_value	= 0;
+	int	swap_cond	= 0;
+	int	prev_value	= 0;
+	int	tval	= 0;
+	int	val_tmp	= 0;
+	int	nmask	= 0;
+	int	a_val	= 0;
+	int	retv	= 0;
+
+	/**
+	 * Can make atomic fetch to be safe in non-cc architecutres
+	 * Has impact on performance
+	 */
+	retv = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default,
+		(void*)next, (void*)&next_value,
+		tval, sizeof(int), my_pe));
+	RUNTIME_CHECK_RC(retv);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+
+	if (next_value == NEXT_MASK) {
+		swap_cond = SET_SIGNAL(my_pe);
+		retv = MCA_ATOMIC_CALL(cswap(oshmem_ctx_default,
+						(void *)tail,
+						(uint64_t *)&(prev_value),
+						OSHMEM_ATOMIC_PTR_2_INT(&swap_cond,
+						sizeof(swap_cond)),
+						OSHMEM_ATOMIC_PTR_2_INT(&val_tmp,
+						sizeof(val_tmp)),
+						sizeof(int),
+						mcs_tail_owner));
+		RUNTIME_CHECK_RC(retv);
+
+		/** I am the tail.. and lock is released */
+		if (prev_value == swap_cond)
+			return;
+		/**
+		 * I am not the tail, another PE maybe racing to acquire lock,
+		 * let them complete setting themselves as our next
+		 */
+		nmask = NEXT_MASK;
+		while(next_value == nmask) {
+			retv = MCA_SPML_CALL(wait((void*)next, SHMEM_CMP_NE,
+					(void*)&nmask, SHMEM_INT));
+			RUNTIME_CHECK_RC(retv);
+			retv = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default,
+						(void*)next,
+						(void*)&next_value,
+						tval, sizeof(int),
+						my_pe));
+			RUNTIME_CHECK_RC(retv);
+		}
+	}
+	/**
+	 * There is a successor release lock to the successor
+	 */
+	a_val = SIGNAL_MASK;
+	retv = MCA_ATOMIC_CALL(add(oshmem_ctx_default,
+				 (void *)next, a_val,
+				 sizeof(a_val), NEXT(next_value)));
+	RUNTIME_CHECK_RC(retv);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+}
+
+int
+_shmem_mcs_test_lock(long *lockp)
+{
+	lock_t *lock = (lock_t *) lockp;
```

Comment:
pls align by `=` for consistency with other vars in this func

Reply:
Done.